sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIO.java

```diff
@@ -1410,31 +1447,34 @@ public void processElement(@Element PubsubMessage message, @Timestamp Instant ti
         }
 
         // NOTE: The record id is always null.
-        output.add(
+        currentTopicOutput.add(
             OutgoingMessage.of(
                 msgBuilder.build(), timestamp.getMillis(), null, message.getTopic()));
-        currentOutputBytes += messageSize;
+        currentOutputBytes.put(pubsubTopic, currentTopicOutputBytes + messageSize);
       }
 
       @FinishBundle
       public void finishBundle() throws IOException {
-        if (!output.isEmpty()) {
-          publish();
-        }
+        publish();
         output = null;
-        currentOutputBytes = 0;
+        currentOutputBytes = null;
         pubsubClient.close();
         pubsubClient = null;
       }
 
       private void publish() throws IOException {
-        PubsubTopic topic = getTopicProvider().get();
+        for (Map.Entry<PubsubTopic, List<OutgoingMessage>> entry : output.entrySet()) {
```

Comment:
use visitor instead of iterator?
forEach(e -> publish(e.getKey(), e.getValue());

Reply:
Unfortunately this does not work for function references that throw exceptions. I would have to wrap a try/catch inside the lambda (i.e. catch the IOException and rethrow as a RuntimeException), which gets ugly.